ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,319 +2,374 @@
+    // DEBUG PRINT
+    if (m_mech_power_print_debug)
+    {
+      printMechanicalPower(m_power_watts, m_mean_mechanical_power);
+    }
 
-  // If the prognoser is ready for a new batch of data, signal the average
-  // mechanical power to be sent in. Otherwise (if the prognoser is currently
-  // processing), discard the current average from this callback.
-  // NOTE: This behavior is not ideal, but since only one batch of data can be
-  //       sent into GSAP's simple prognoser per 0.5Hz prediction, message data
-  //       has to be discarded. Once the pack model with its asynchronous
-  //       prognosers is implemented, all data can be sent in regardless of the
-  //       prediction speed of GSAP.
-  if (!m_processing_power_batch)
-  {
-    m_mechanical_power_to_be_processed = m_unprocessed_mechanical_power;
-    m_unprocessed_mechanical_power = 0.0;   // reset the accumulator(?)
-    m_trigger_processing_new_power_batch = true;
-  }
-}
+    // Unlock the power system.
+    m_processing_power_batch = false;
 
-double PowerSystemNode::generateTemperatureEstimate()
-{
-  return m_temperature_dist(m_random_generator);
-}
+    // Check if the prediction handlers have EoD predictions ready,
+    // and update EoD_events if so.
+    for (int i = 0; i < NUM_MODELS; i++)
+    {
+      if (handlers[i]->getStatus())
+      {
+        m_EoD_events[i] = handlers[i]->getEoD();
+      }
+    }
 
-double PowerSystemNode::generateVoltageEstimate()
-{
-  // Create voltage estimate with pseudorandom noise generator - needs to decrease over time
-  double timelapse = duration<double>(system_clock::now() - m_init_time).count();               // [s]
-  double min_V = m_base_voltage + (m_battery_lifetime - timelapse) / m_battery_lifetime * 0.8;  // [V]
-  double max_V = min_V + m_voltage_range;                                                       // [V]
+    // Publish the predictions in EoD_events.
+    publishPredictions();
 
-  // If voltage limits dip below baseline, set to baseline values
-  if (min_V < m_base_voltage)
-  {
-    min_V = m_base_voltage;
+    // DEBUG PRINT
+    if (m_outputs_print_debug)
+    {
+      ROS_INFO_STREAM("CURRENT MODEL DATA:");
+      for (int i = 0; i < NUM_MODELS; i++)
+      {
+        printPrognoserOutputs(m_EoD_events[i].remaining_useful_life,
+                              m_EoD_events[i].state_of_charge,
+                              m_EoD_events[i].battery_temperature,
+                              i);
+      }
+    }
+
+    // DEBUG PRINT FORMATTING
+    if (m_print_debug && !(m_power_models[0].input_info.timestamp <= 0))
+    {
+      std::cout << std::endl;
+    }
+
+    // Sleep for any remaining time in the loop that would cause it to
+    // complete before the set rate of GSAP.
+    // NOTE: Currently there is no catch for if the loop takes longer than the
+    // specified ROS rate. If this occurs, the simulation will lag behind real
+    // time.
```

Comment:
Correct, this is basically impossible now. When this comment was written though, the single-cell model was in place (to my memory), which came much closer to surpassing 2 seconds per prediction. In fact, last year during my first attempts to implement this, it would take upwards of 30 seconds per batch of predictions, causing the time to increment by a mere 4 or 6s over a full 60 seconds.

I can probably remove this comment now, or at the very least clarify that such a situation is highly unlikely unless something within GSAP breaks mid-simulation.

Reply:
This is handy
https://docs.ros.org/en/noetic/api/rostime/html/classros_1_1Rate.html#ae5664d27cda1b17a103347560259e945

It turns out the `roscpp::Rate::sleep` returns `false` if the "rate was not met" for the cycle. I think it's good to handle the `false` return case with a ROS warning. I don't think the comment is necessary anymore if we do this.